tests/Unit/API/Site/Controllers/MerchantCenter/AttributeMappingRulesControllerTest.php

```diff
@@ -0,0 +1,194 @@
+	}
+
+	public function test_create_rule_route_fails() {
+		$this->attribute_mapping_helper->expects( $this->any() )
+			->method( 'get_attributes' )->willReturn(
+				self::TEST_ATTRIBUTES
+			);
+
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'insert' )->willReturn( 0 );
+
+		$response = $this->do_request( self::ROUTE_RULES, 'post', self::TEST_RULE );
+		$this->assertEquals( 400, $response->get_status() );
+	}
+
+	public function test_update_rule_route_fails() {
+		$this->attribute_mapping_helper->expects( $this->any() )
+			->method( 'get_attributes' )->willReturn(
+				self::TEST_ATTRIBUTES
+			);
+
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'update' )->willReturn( 0 );
+
+		$response = $this->do_request( self::ROUTE_RULE, 'post', self::TEST_RULE );
+		$this->assertEquals( 400, $response->get_status() );
+	}
+
+	private function validate_post_route( $route, $method ) {
+		$rule_with_id = array_merge( [ 'id' => self::TEST_RULE_ID ], self::TEST_RULE, [ 'categories' => null ] );
+
+		$this->attribute_mapping_helper->expects( $this->any() )
+			->method( 'get_attributes' )->willReturn(
+				self::TEST_ATTRIBUTES
+			);
+
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( $method )->willReturn( 1 );
+
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'get_rule' )->willReturn( $rule_with_id );
+
+		// insert works
+		$response = $this->do_request( $route, 'post', self::TEST_RULE );
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( $rule_with_id, $response->get_data() );
+
+		// not working without rule.attribute param
+		$response = $this->do_request(
+			$route,
+			'post',
+			[
+				'source'                  => 'x',
+				'category_condition_type' => 'ALL',
+			]
+		);
+		$this->assertEquals( 400, $response->get_status() );
+
+		// not working without rule.source param
+		$response = $this->do_request(
+			$route,
+			'post',
+			[
+				'attribute'               => 'adult',
+				'category_condition_type' => 'ALL',
+			]
+		);
+		$this->assertEquals( 400, $response->get_status() );
+
+		// not working without rule.category_condition_type param
+		$response = $this->do_request(
+			$route,
+			'post',
+			[
+				'attribute' => 'adult',
+				'source'    => 'test',
+			]
+		);
+		$this->assertEquals( 400, $response->get_status() );
+
+	}
+
+	public function test_delete_rule_route() {
+		$data = [ 'id' => self::TEST_RULE_ID ];
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'delete' )->willReturn( 1 );
+
+		// delete works
+		$response = $this->do_request( self::ROUTE_RULE, 'delete' );
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( $data, $response->get_data() );
+	}
+
+	public function test_delete_rule_route_error() {
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'delete' )->willReturn( 0 );
```

Comment:
According to the definition of the delete function it `throws InvalidQuery When there is an error deleting data`
Have we formatted the delete route incorrectly? We should only have to catch the errors and never need to check for a `0` value being returned.

Reply:
It returns 0 when no rows where deleted. Which is not an error. We are catching error as well.

Reply:
I guess I should clarify. In the `wpdb::delete` function and the `Query::delete` function not deleting a row (maybe because it was previously deleted) is not considered an error. It will only throw an error if it encounters a problem like connecting to the database or the SQL table is missing or something related.

However in the `AttributeMappingRulesController` we explicitly make the delete return an error if it didn't actually delete a row. That's different then the behaviour in for example deleting a shipping rate or time. If that's intentional then that's fine, I was mainly pointing out the difference.